### ISSUE TITLE
[M1-03] On-policy REINFORCE baseline

### ISSUE THREAD

**Issue by anton-pershin** [dynolab/off-policy-for-toy-llms#1]
## 1. Description

An on-policy REINFORCE training harness for the contextual-bandit NLI task, together with the canonical definition of that task's interface layer. The harness loads the synthetic NLI dataset, renders each example into a prompt, samples a completion from a small decoder-only LLM, parses a label from the completion, computes a deterministic reward, and updates the policy by policy gradient.

Because it is the first implementation in the project, this spec also fixes the bandit formalization that later specs depend on:

- **context / action / reward** — context is a `(premise, hypothesis)` pair from the dataset; the action is the full generated completion (not the parsed label), so that sequence-level importance ratios are well-defined for tasks 04–05; the reward is deterministic in the parsed label and `gold_label`
- **prompt construction** — a deterministic, versioned template mapping a dataset example to a prompt string
- **parsing** — a deterministic function from completion text to a canonical label or parse-failure, over the label set
- **per-sample record schema** — the fields logged for every interaction, including behaviour-policy bookkeeping

Delivered scope is the on-policy path only: one policy, no replay buffer, no policy lag, no off-policy corrections. Components are built with explicit extension seams — a pluggable loss objective, a batch-source abstraction, and behaviour-policy bookkeeping written from day one — so we could  extend this harness in future rather than replace it.

## 2. Motivation

Every later M1 task builds on this harness, and every off-policy result is measured against the baseline it produces. Four things must be established before any off-policiness knob is meaningful:

1. **The task interface exists and is canonical.** Prompt, parser, reward, and record schema are currently undefined in any current document. Off-policy corrections are defined over the action's log-probability, so the action must be pinned down before they can be implemented.
2. **The pipeline is correct.** Reward parsing, sequence log-probability computation, and gradient flow must be verified in the simplest possible setting. Debugging them jointly with a replay buffer and importance weights is far harder — an incorrect `logp` and a genuine off-policy instability present similarly.
3. **The task is learnable.** If the model cannot exceed the 33% chance floor under ideal on-policy conditions, there is no signal for off-policiness to degrade and the testbed is invalid.
4. **The baseline exists.** RQ1's "effect on stability, sample efficiency, and final accuracy" is defined as a difference from on-policy training; without this reference curve no off-policy measurement is interpretable.

### 3.1 Functional requirements

**Task interface (canonical definitions)**

- **FR1 — Prompt construction.** A deterministic function mapping a dataset example to a prompt string, using a versioned template. The template identifier must be recorded per run. The prompt must present `premise` and `hypothesis` and request one of the three canonical labels.
- **FR2 — Label parsing.** A deterministic function `parse(completion_text) -> (label | None, parse_success)` over the canonical label set (`entailment`, `neutral`, `contradiction`). Parsing must be case-insensitive, must tolerate leading whitespace/punctuation, and must accept completions containing text beyond the label. It must never raise on arbitrary model output.
- **FR3 — Reward.** A deterministic function of the parsed label and `gold_label`: reward 1 for a correct parsed label, 0 otherwise; a parse failure yields reward 0. Reward must depend on nothing else (no length or format bonuses) so that the reward channel remains noise-free.
- **FR4 — Action definition.** The action is the full generated completion token sequence. All log-probabilities are sequence-level sums over generated tokens only, excluding prompt tokens.

**Data and generation**

- **FR5 — Dataset loading.** Load JSONL splits conforming to the data spec 02 schema, exposing `train` and `eval` splits. Must validate required fields on load and fail loudly on schema violations. Must support filtering by `hop_difficulty`, `distractor_difficulty`, and `background_knowledge` so the difficulty mix of a run is explicit and controllable.
- **FR6 — Batched sampling.** Sample completions for a batch of prompts from the current policy with configurable decoding parameters (temperature, top-p, max new tokens). The decoding parameters actually used must be recorded per sample.
- **FR7 — Behaviour log-probability.** For each sampled completion, compute and store the sequence log-probability under the policy that generated it, at generation time.

**Training**

- **FR8 — Policy gradient update.** Implement the REINFORCE objective with a configurable baseline for variance reduction (at minimum: no baseline, and batch-mean reward). The loss must be exposed through a pluggable objective interface so tasks 04–05 can add IS-clipped and KL-regularized variants without modifying the training loop.
- **FR9 — Target log-probability recomputation.** Recompute the sequence log-probability of a stored completion under the current policy — the quantity gradients flow through, and the numerator of the importance ratio in task 04. On-policy this is computed on freshly sampled data; the interface must not assume the sampling policy and the updating policy are identical.
- **FR10 — Batch source abstraction.** The training loop must obtain batches through an abstraction whose on-policy implementation samples fresh data each step, so task 03 can substitute a replay-buffer-backed source without touching the loop.
- **FR11 — Evaluation.** Periodic evaluation on the held-out split, reporting accuracy and mean reward. Evaluation decoding settings must be configurable independently of training (in particular, greedy decoding must be available).

**Logging and reproducibility**

- **FR12 — Per-sample records.** Every interaction is recorded with: example id, prompt text, gold label, completion text and token ids, parsed label, parse success, reward, behaviour log-probability, behaviour policy identifier, learner step, and the decoding parameters used. The schema must accommodate the fields tasks 03–05 require (staleness, importance weights) without a breaking change.
- **FR13 — Run metrics.** Log per training step: mean reward, running accuracy, parse success rate, loss, gradient norm, policy entropy, and mean/std of the sequence log-probability. Log per evaluation: held-out accuracy and mean reward. Metrics must be persisted in a machine-readable form for later plotting.
- **FR14 — Configuration.** All experiment parameters supplied through configuration files rather than code edits or hardcoded defaults, with the fully resolved config saved alongside the run's outputs.
- **FR15 — Seeding.** A single run seed controls model init, data ordering, and sampling; re-running with the same seed and config reproduces the same metrics.

### 3.2 Non-functional requirements

- **NFR1 — Single-GPU budget.** Must run end-to-end on one consumer GPU (RTX 3090/4090 class, 24 GB) per constitution §7, including a base model, sampling, and gradient computation. Batch size and sequence length defaults must fit this budget.
- **NFR2 — Baseline run turnaround.** A full baseline run should complete within a few hours, so that later sweeps over off-policiness knobs (dozens to hundreds of runs) remain feasible within the milestone.
- **NFR3 — Extensibility.** Adding an off-policy objective (task 04) or a replay-backed batch source (task 03) must require implementing the corresponding interface only, with no changes to the training loop, task interface, or record schema.
- **NFR4 — Determinism.** Given a fixed seed and config, runs are reproducible up to documented GPU non-determinism; any deliberately non-deterministic component must be identified.
- **NFR5 — Model-agnosticism.** The base model is a configuration value. Nothing in the harness may assume a specific checkpoint or tokenizer, so a second model size can be swapped in without code changes.
- **NFR6 — Testability.** Task interface functions (prompt, parse, reward) and log-probability computation must be unit-testable without a GPU; a tiny model must be usable in tests for the full loop.

### 4.1 High-level design

```mermaid
flowchart TB
    subgraph cfg[Configuration]
        CFG[Run config<br/>model, data, sampling,<br/>objective, optimizer, seed]
    end

    subgraph task[Task interface - canonical]
        DS[NLIDataset<br/>JSONL loader + difficulty filter]
        PB[PromptBuilder<br/>versioned template]
        PR[LabelParser<br/>completion to label]
        RW[RewardFn<br/>parsed label vs gold]
    end

    subgraph policy[Policy]
        PM[Policy<br/>model + tokenizer]
        GEN[sample<br/>batched generation + logp]
        LP[sequence_logp<br/>recompute under current params]
    end

    subgraph loop[Training loop]
        BS[BatchSource<br/>on-policy: sample fresh]
        OBJ[Objective<br/>REINFORCE + baseline]
        OPT[Optimizer step]
        EV[Evaluator<br/>held-out accuracy]
    end

    subgraph out[Outputs]
        REC[Per-sample records]
        MET[Run metrics]
        RCFG[Resolved config]
    end

    CFG --> DS & PM & BS & OBJ & EV
    DS --> PB --> GEN
    PM --> GEN & LP
    BS --> GEN
    GEN --> PR --> RW --> BS
    BS -->|batch| OBJ
    LP --> OBJ
    OBJ --> OPT --> PM
    BS --> REC
    OBJ --> MET
    EV --> MET
    CFG --> RCFG
    PB & PR & RW -.-> EV
```

Control flow per training step: `BatchSource` produces a batch (on-policy: draw contexts from the dataset, build prompts, sample completions, parse, reward, record); the `Objective` recomputes target log-probabilities under current parameters and returns a scalar loss; the optimizer updates the policy. Periodically the `Evaluator` runs the same task interface on the held-out split with independent decoding settings.

### 4.2 Core components

**Task interface**

- **`NLIDataset`** — loads JSONL splits, validates required fields on load and fails loudly on violations, exposes `train`/`eval` splits, and supports filtering by `hop_difficulty`, `distractor_difficulty` and `background_knowledge`. Owns nothing about prompts or models. *(FR5)*
- **`PromptBuilder`** — deterministic `example -> prompt string` under a named, versioned template; exposes its template identifier for logging. *(FR1)*
- **`LabelParser`** — deterministic `completion text -> (label | None, parse_success)` over the three canonical labels; total (never raises), case-insensitive, tolerant of leading punctuation and trailing text. *(FR2)*
- **`RewardFn`** — deterministic `(parsed label, parse_success, gold label) -> {0, 1}`; depends on nothing else. *(FR3)*

**Policy**

- **`Policy`** — wraps a configured causal LM and its tokenizer; the base checkpoint is a config value with no assumptions elsewhere in the harness. Exposes two operations: `sample(prompts, decoding params) -> completions + behaviour log-probabilities` (no gradient) and `sequence_logp(prompts, completions) -> log-probabilities` (differentiable). Both compute sequence-level sums over generated tokens only, sharing one masking implementation so the two paths cannot diverge. Also exposes a `policy_id` identifying the current parameter state, recorded per sample. *(FR4, FR6, FR7, FR9, NFR5)*

**Training loop**

- **`BatchSource`** (interface) — yields training batches. `OnPolicyBatchSource` draws contexts, builds prompts, samples, parses, rewards, and emits per-sample records. *(FR10, FR12)*
- **`Objective`** (interface) — `(batch, policy) -> (loss, diagnostics)`. `ReinforceObjective` implements the policy gradient with a configurable baseline (`none` or `batch_mean`) and reports diagnostics (loss, entropy, log-probability statistics). *(FR8, NFR3)*
- **`Trainer`** — owns the step loop: pull batch, compute loss, clip and apply gradients, log metrics, trigger periodic evaluation and checkpointing. Contains no task-specific or objective-specific logic. *(FR13)*
- **`Evaluator`** — runs the task interface over the held-out split with its own decoding configuration (greedy available), reporting accuracy, mean reward and parse success rate. *(FR11)*

**Infrastructure**

- **`RecordWriter`** — appends per-sample records in the FR12 schema to a machine-readable file. *(FR12)*
- **`MetricsLogger`** — persists per-step and per-eval metrics in a machine-readable form for later plotting; the single place metrics are written, so adding off-policy diagnostics later touches one component. *(FR13)*
- **Config and entry point** — a config-driven run script; the fully resolved config is saved with the run outputs and is sufficient to reproduce the run. Seeding is centralized so one seed determines model init, data order and sampling. *(FR14, FR15, NFR4)*
